src/cmap/auth/gssapi.ts

```diff
@@ -30,60 +51,161 @@ export class GSSAPI extends AuthProvider {
-              if (err) return callback(err, false);
+                      externalCommand(
+                        {
+                          saslContinue: 1,
+                          conversationId: doc.conversationId,
+                          payload
+                        },
+                        (err, result) => {
+                          if (err) return callback(err);
 
-              const doc = result.result;
-              authProcess.transition(doc.payload, (err, payload) => {
-                if (err) return callback(err, false);
-                const command = {
-                  saslContinue: 1,
-                  conversationId: doc.conversationId,
-                  payload
-                };
-
-                connection.command('$external.$cmd', command, (err, result) => {
-                  if (err) return callback(err, false);
-
-                  const response = result.result;
-                  authProcess.transition(null, err => {
-                    if (err) return callback(err);
-                    callback(undefined, response);
-                  });
-                });
+                          callback(undefined, result.result);
+                        }
+                      );
+                    });
+                  }
+                );
               });
-            });
-          });
+            }
+          );
         });
-      });
+      }
+    );
+  }
+}
+
+function initialize(
+  context: Omit<GSSAPIContext, 'client'>,
+  callback: (err?: Error, context?: GSSAPIContext) => void
+): void {
+  // Canonicialize host name if needed
+  function performGssapiCanonicalizeHostName(
+    canonicalizeHostName: boolean,
+    host: string,
+    callback: Function
+  ) {
+    if (!canonicalizeHostName) return callback();
+
+    // Attempt to resolve the host name
+    dns.resolveCname(host, (err: Error | null, r: string[]) => {
+      if (err) return callback(err);
+
+      // Get the first resolve host id
+      if (Array.isArray(r) && r.length > 0) {
+        context.host = r[0];
+      }
+
+      callback();
     });
   }
+
+  // Canonicialize host name if needed
+  performGssapiCanonicalizeHostName(context.canonicalizeHostName, context.host, (err: Error) => {
+    if (err) return callback(err);
+
+    const initOptions = {};
+    if (context.password != null) {
+      Object.assign(initOptions, { user: context.username, password: context.password });
+    }
+
+    Kerberos.initializeClient(
+      `${context.serviceName}${process.platform === 'win32' ? '/' : '@'}${context.host}`,
+      initOptions,
+      (err: string, client: KerberosClient): void => {
+        if (err) return callback(new Error(err));
+        const c: GSSAPIContext = Object.assign({ client }, context);
+        callback(undefined, c);
+      }
+    );
+  });
+}
+
+function negotiate(context: GSSAPIContext, payload: string, callback: Callback<string>) {
+  context.client.step(payload, (err, response) => {
```

Comment:
nit: pluck out `client` and `retries` here. I don't think `retries` should be mutated on the `context` object, rather this method should take in a constant and use a local variable to decrement during retries. Let's try to be as "immutable" as possible.

In fact, I feel like maybe this would just be better suited as a util function, we effectively want [async.retry](https://caolan.github.io/async/v3/docs.html#retry)